### PR TITLE
Intergrate trocador

### LINKED ISSUE #253: add trocador as part of the exchanges available to users
this issue depends on the work done in this PR https://github.com/crypto-power/instantswap/pull/7

Image

Website: https://trocador.app/

### PR DESCRIPTION
**Pull request by vibros68**
closes #253 

This diff adds support for trocador instant exchange 
Screenshot from 2023-11-17 22-44-50

### CODE REVIEW

File: ui/page/exchange/create_order_page.go [hunk 1]
```diff
@@ -1127,6 +1140,11 @@ func (pg *CreateOrderPage) getExchangeRateInfo() error {
 		pg.fetchingRate = false
 		return err
 	}
+	fmt.Printf("%+v \n", params)
```
Comment (dreacot) [inline ui/page/exchange/create_order_page.go:1143]: remove print statement

**Review by dreacot — CHANGES_REQUESTED** [review]
trying to swap ltc and dcr throws an error
Untitled

**Review by dreacot — COMMENTED** [review]
the conversion factor is wrong, 1 DCR  should not give 2.12 LTC

Screenshot from 2023-11-17 10-55-08

File: ui/page/exchange/create_order_page.go [hunk 2]
```diff
@@ -1116,9 +1127,11 @@ func (pg *CreateOrderPage) getExchangeRateInfo() error {
...
 	toCur := pg.toCurrency.String()
 	params := api.ExchangeRateRequest{
-		From:   fromCur,
-		To:     toCur,
-		Amount: libwallet.DefaultRateRequestAmount, // amount needs to be greater than 0 to get the exchange rate
+		From:        fromCur,
+		FromNetwork: getNetwork(fromCur, pg.instantCurrencies),
+		To:          toCur,
+		ToNetwork:   getNetwork(toCur, pg.instantCurrencies),
+		Amount:      libwallet.RateRequest(fromCur), // amount needs to be greater than 0 to get the exchange rate
```
Comment (dreacot) [inline ui/page/exchange/create_order_page.go:1134]: why are we overriding the amount inputted by the user?

**Comment by Sirmorrison** [comment]
https://github.com/crypto-power/cryptopower/issues/243#issuecomment-1818447077

File: ui/page/exchange/create_order_page.go (hunk 2 above)
Comment (dreacot) [reply ui/page/exchange/create_order_page.go:1134]: oh i see this is just ti get the exchange rate

Comment (vibros68) [reply ui/page/exchange/create_order_page.go:1134]: Yes

**Comment by dreacot** [comment]
after placing a trade with trocador, i get error message "Unsupported asset type" in my recent order list

Screenshot from 2023-11-24 09-31-29

**Comment by dreacot** [comment]
clicking to open the trade, i get the following error

```
2023-11-24 09:32:43.377 [ERR] SHWL: instantSwap.GetOrderInfo: json: cannot unmarshal array into Go value of type trocador.Trade
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x168 pc=0x13a4b02]

goroutine 24 [running]:
github.com/crypto-power/cryptopower/ui/page/exchange.(*OrderDetailsPage).layout.func1.1({{{0x0, 0x399}, {0x226, 0x399}}, {0x3f800000, 0x3f800000}, {0x1ef1be0, 0xc0038780c8}, {0xc15051ead6a83efa, 0x23eb3e0b5a, ...}, ...})
        /home/kennedy/Projects/src/github.com/cryptopower/cryptopower/ui/page/exchange/order_details_page.go:249 +0xa2
gioui.org/layout.Flex.Layout({0x19?, 0x0?, 0x0?, 0x0?}, {{{0x0, 0x399}, {0x226, 0x399}}, {0x3f800000, 0x3f800000}, ...}, ...)
        /home/kennedy/Projects/pkg/mod/gioui.org@v0.1.0/layout/flex.go:99 +0x2d5
github.com/crypto-power/cryptopower/ui/cryptomaterial.LinearLayout.Layout.func1.1.1.2.1.1.1({{{0x0, 0x399}, {0x226, 0x399}}, {0x3f800000, 0x3f800000}, {0x1ef1be0, 0xc0038780c8}, {0xc15051ead6a83efa, 0x23eb3e0b5a, ...}, ...})
        /home/kennedy/Projects/src/github.com/cryptopower/cryptopower/ui/cryptomaterial/linearlayout.go:88 +0x85
gioui.org/layout.Direction.Layout(0x7, {{{0x0, 0x399}, {0x226, 0x399}}, {0x3f800000, 0x3f800000}, {0x1ef1be0, 0xc0038780c8}, {0xc15051ead6a83efa, ...}, ...}, ...)
        /home/kennedy/Projects/pkg/mod/gioui.org@v0.1.0/layout/layout.go:196 +0x162
github.com/crypto-power/cryptopower/ui/cryptomaterial.LinearLayout.Layout.func1.1.1.2.1.1({{{0x226, 0x399}, {0x226, 0x399}}, {0x3f800000, 0x3f800000}, {0x1ef1be0, 0xc0038780c8}, {0xc15051ead6a83efa, 0x23eb3e0b5a, ...}, ...})
        /home/kennedy/Projects/src/github.com/cryptopower/cryptopower/ui/cryptomaterial/linearlayout.go:87 +0xe5
gioui.org/layout.Inset.Layout({0x2f9b5b0?, 0xc0?, 0x6b3f57?, 0x0?}, {{{0x226, 0x399}, {0x226, 0x399}}, {0x3f800000, 0x3f800000}, ...}, ...)
        /home/kennedy/Projects/pkg/mod/gioui.org@v0.1.0/layout/layout.go:169 +0x39d
github.com/crypto-power/cryptopower/ui/cryptomaterial.LinearLayout.Layout.func1.1.1.2.1({{{0x226, 0x399}, {0x226, 0x399}}, {0x3f800000, 0x3f800000}, {0x1ef1be0, 0xc0038780c8}, {0xc15051ead6a83efa, 0x23eb3e0b5a, ...}, ...})
        /home/kennedy/Projects/src/github.com/cryptopower/cryptopower/ui/cryptomaterial/linearlayout.go:85 +0x109
github.com/crypto-power/cryptopower/ui/cryptomaterial.Border.Layout({{0x0, 0x0, 0x0, 0x0}, {0x0, 0x0, 0x0, 0x0}, 0x0}, {{{0x226, ...}, ...}, ...}, ...)
        /home/kennedy/Projects/src/github.com/cryptopower/cryptopower/ui/cryptomaterial/border.go:21 +0x98
github.com/crypto-power/cryptopower/ui/cryptomaterial.LinearLayout.Layout.func1.1.1.2({{{0x226, 0x399}, {0x226, 0x399}}, {0x3f800000, 0x3f800000}, {0x1ef1be0, 0xc0038780c8}, {0xc15051ead6a83efa, 0x23eb3e0b5a, ...}, ...})
        /home/kennedy/Projects/src/github.com/cryptopower/cryptopower/ui/cryptomaterial/linearlayout.go:83 +0x1d6
gioui.org/layout.Stack.Layout({0xd?}, {{{0x0, 0x0}, {0x550, 0x399}}, {0x3f800000, 0x3f800000}, {0x1ef1be0, 0xc0038780c8}, {0xc15051ead6a83efa, ...}, ...}, ...)
        /home/kennedy/Projects/pkg/mod/gioui.org@v0.1.0/layout/stack.go:60 +0x195
github.com/crypto-power/cryptopower/ui/cryptomaterial.LinearLayout.Layout.func1.1.1({{{0x0, 0x0}, {0x550, 0x399}}, {0x3f800000, 0x3f800000}, {0x1ef1be0, 0xc0038780c8}, {0xc15051ead6a83efa, 0x23eb3e0b5a, ...}, ...})
        /home/kennedy/Projects/src/github.com/cryptopower/cryptopower/ui/cryptomaterial/linearlayout.go:46 +0x2aa
github.com/crypto-power/cryptopower/ui/cryptomaterial.LinearLayout.Layout.func1.1({{{0x0, 0x0}, {0x550, 0x399}}, {0x3f800000, 0x3f800000}, {0x1ef1be0, 0xc0038780c8}, {0xc15051ead6a83efa, 0x23eb3e0b5a, ...}, ...})
        /home/kennedy/Projects/src/github.com/cryptopower/cryptopower/ui/cryptomaterial/linearlayout.go:105 +0x238
gioui.org/layout.Inset.Layout({0x0?, 0x0?, 0x0?, 0x0?}, {{{0x0, 0x0}, {0x550, 0x399}}, {0x3f800000, 0x3f800000}, ...}, ...)
        /home/kennedy/Projects/pkg/mod/gioui.org@v0.1.0/layout/layout.go:169 +0x39d
github.com/crypto-power/cryptopower/ui/cryptomaterial.LinearLayout.Layout.func1({{{0x0, 0x0}, {0x550, 0x3b7}}, {0x3f800000, 0x3f800000}, {0x1ef1be0, 0xc0038780c8}, {0xc15051ead6a83efa, 0x23eb3e0b5a, ...}, ...})
        /home/kennedy/Projects/src/github.com/cryptopower/cryptopower/ui/cryptomaterial/linearlayout.go:44 +0x109
gioui.org/layout.Direction.Layout(0x7, {{{0x0, 0x0}, {0x550, 0x3b7}}, {0x3f800000, 0x3f800000}, {0x1ef1be0, 0xc0038780c8}, {0xc15051ead6a83efa, ...}, ...}, ...)
        /home/kennedy/Projects/pkg/mod/gioui.org@v0.1.0/layout/layout.go:196 +0x162
github.com/crypto-power/cryptopower/ui/cryptomaterial.LinearLayout.Layout({0x226, 0xfffffffffffffffe, 0x0, {0x0, 0x0, 0x0, 0x0}, 0x0, {{0x0, 0x0, ...}, ...}, ...}, ...)
        /home/kennedy/Projects/src/github.com/cryptopower/cryptopower/ui/cryptomaterial/linearlayout.go:42 +0xe5
github.com/crypto-power/cryptopower/ui/cryptomaterial.LinearLayout.Layout2(...)
        /home/kennedy/Projects/src/github.com/cryptopower/cryptopower/ui/cryptomaterial/linearlayout.go:37
github.com/crypto-power/cryptopower/ui/page/exchange.(*OrderDetailsPage).layout.func1({{{0x0, 0x0}, {0x550, 0x3b7}}, {0x3f800000, 0x3f800000}, {0x1ef1be0, 0xc0038780c8}, {0xc15051ead6a83efa, 0x23eb3e0b5a, ...}, ...})
        /home/kennedy/Projects/src/github.com/cryptopower/cryptopower/ui/page/exchange/order_details_page.go:155 +0x22a
gioui.org/layout.Flex.Layout({0x15?, 0x0?, 0x0?, 0x0?}, {{{0x0, 0x0}, {0x550, 0x3b7}}, {0x3f800000, 0x3f800000}, ...}, ...)
        /home/kennedy/Projects/pkg/mod/gioui.org@v0.1.0/layout/flex.go:99 +0x2d5
github.com/crypto-power/cryptopower/ui/cryptomaterial.LinearLayout.Layout.func1.1.1.2.1.1.1({{{0x0, 0x0}, {0x550, 0x3b7}}, {0x3f800000, 0x3f800000}, {0x1ef1be0, 0xc0038780c8}, {0xc15051ead6a83efa, 0x23eb3e0b5a, ...}, ...})
        /home/kennedy/Projects/src/github.com/cryptopower/cryptopower/ui/cryptomaterial/linearlayout.go:88 +0x85
gioui.org/layout.Direction.Layout(0x8, {{{0x0, 0x0}, {0x550, 0x3b7}}, {0x3f800000, 0x3f800000}, {0x1ef1be0, 0xc0038780c8}, {0xc15051ead6a83efa, ...}, ...}, ...)
        /home/kennedy/Projects/pkg/mod/gioui.org@v0.1.0/layout/layout.go:196 +0x162
github.com/crypto-power/cryptopower/ui/cryptomaterial.LinearLayout.Layout.func1.1.1.2.1.1({{{0x550, 0x3b7}, {0x550, 0x3b7}}, {0x3f800000, 0x3f800000}, {0x1ef1be0, 0xc0038780c8}, {0xc15051ead6a83efa, 0x23eb3e0b5a, ...}, ...})
        /home/kennedy/Projects/src/github.com/cryptopower/cryptopower/ui/cryptomaterial/linearlayout.go:87 +0xe5
gioui.org/layout.Inset.Layout({0xb?, 0x0?, 0x3eb?, 0x0?}, {{{0x550, 0x3b7}, {0x550, 0x3b7}}, {0x3f800000, 0x3f800000}, ...}, ...)
        /home/kennedy/Projects/pkg/mod/gioui.org@v0.1.0/layout/layout.go:169 +0x39d
github.com/crypto-power/cryptopower/ui/cryptomaterial.LinearLayout.Layout.func1.1.1.2.1({{{0x550, 0x3b7}, {0x550, 0x3b7}}, {0x3f800000, 0x3f800000}, {0x1ef1be0, 0xc0038780c8}, {0xc15051ead6a83efa, 0x23eb3e0b5a, ...}, ...})
        /home/kennedy/Projects/src/github.com/cryptopower/cryptopower/ui/cryptomaterial/linearlayout.go:85 +0x109
github.com/crypto-power/cryptopower/ui/cryptomaterial.Border.Layout({{0x0, 0x0, 0x0, 0x0}, {0x0, 0x0, 0x0, 0x0}, 0x0}, {{{0x550, ...}, ...}, ...}, ...)
        /home/kennedy/Projects/src/github.com/cryptopower/cryptopower/ui/cryptomaterial/border.go:21 +0x98
github.com/crypto-power/cryptopower/ui/cryptomaterial.LinearLayout.Layout.func1.1.1.2({{{0x550, 0x3b7}, {0x550, 0x3b7}}, {0x3f800000, 0x3f800000}, {0x1ef1be0, 0xc0038780c8}, {0xc15051ead6a83efa, 0x23eb3e0b5a, ...}, ...})
        /home/kennedy/Projects/src/github.com/cryptopower/cryptopower/ui/cryptomaterial/linearlayout.go:83 +0x1d6
gioui.org/layout.Stack.Layout({0x20?}, {{{0x0, 0x0}, {0x550, 0x3b7}}, {0x3f800000, 0x3f800000}, {0x1ef1be0, 0xc0038780c8}, {0xc15051ead6a83efa, ...}, ...}, ...)
        /home/kennedy/Projects/pkg/mod/gioui.org@v0.1.0/layout/stack.go:60 +0x195
github.com/crypto-power/cryptopower/ui/cryptomaterial.LinearLayout.Layout.func1.1.1({{{0x0, 0x0}, {0x550, 0x3b7}}, {0x3f800000, 0x3f800000}, {0x1ef1be0, 0xc0038780c8}, {0xc15051ead6a83efa, 0x23eb3e0b5a, ...}, ...})
        /home/kennedy/Projects/src/github.com/cryptopower/cryptopower/ui/cryptomaterial/linearlayout.go:46 +0x2aa
github.com/crypto-power/cryptopower/ui/cryptomaterial.LinearLayout.Layout.func1.1({{{0x0, 0x0}, {0x550, 0x3b7}}, {0x3f800000, 0x3f800000}, {0x1ef1be0, 0xc0038780c8}, {0xc15051ead6a83efa, 0x23eb3e0b5a, ...}, ...})
        /home/kennedy/Projects/src/github.com/cryptopower/cryptopower/ui/cryptomaterial/linearlayout.go:105 +0x238
gioui.org/layout.Inset.Layout({0x1?, 0x3f800000?, 0x9823ff0?, 0xc0?}, {{{0x0, 0x0}, {0x550, 0x3b7}}, {0x3f800000, 0x3f800000}, ...}, ...)
        /home/kennedy/Projects/pkg/mod/gioui.org@v0.1.0/layout/layout.go:169 +0x39d
github.com/crypto-power/cryptopower/ui/cryptomaterial.LinearLayout.Layout.func1({{{0x0, 0x0}, {0x550, 0x3b7}}, {0x3f800000, 0x3f800000}, {0x1ef1be0, 0xc0038780c8}, {0xc15051ead6a83efa, 0x23eb3e0b5a, ...}, ...})
        /home/kennedy/Projects/src/github.com/cryptopower/cryptopower/ui/cryptomaterial/linearlayout.go:44 +0x109
gioui.org/layout.Direction.Layout(0x8, {{{0x0, 0x0}, {0x550, 0x3b7}}, {0x3f800000, 0x3f800000}, {0x1ef1be0, 0xc0038780c8}, {0xc15051ead6a83efa, ...}, ...}, ...)
        /home/kennedy/Projects/pkg/mod/gioui.org@v0.1.0/layout/layout.go:196 +0x162
github.com/crypto-power/cryptopower/ui/cryptomaterial.LinearLayout.Layout({0xfffffffffffffffe, 0xfffffffffffffffe, 0x0, {0x0, 0x0, 0x0, 0x0}, 0x0, {{0x0, 0x0, ...}, ...}, ...}, ...)
        /home/kennedy/Projects/src/github.com/cryptopower/cryptopower/ui/cryptomaterial/linearlayout.go:42 +0xe5
github.com/crypto-power/cryptopower/ui/cryptomaterial.LinearLayout.Layout2(...)
        /home/kennedy/Projects/src/github.com/cryptopower/cryptopower/ui/cryptomaterial/linearlayout.go:37
github.com/crypto-power/cryptopower/ui/page/exchange.(*OrderDetailsPage).layout(0xc003b6d500, {{{0x0, 0x0}, {0x550, 0x3b7}}, {0x3f800000, 0x3f800000}, {0x1ef1be0, 0xc0038780c8}, {0xc15051ead6a83efa, ...}, ...})
        /home/kennedy/Projects/src/github.com/cryptopower/cryptopower/ui/page/exchange/order_details_page.go:147 +0x1aa
gioui.org/layout.Flex.Layout({0x0?, 0x0?, 0x0?, 0x0?}, {{{0x0, 0x0}, {0x550, 0x3e7}}, {0x3f800000, 0x3f800000}, ...}, ...)
        /home/kennedy/Projects/pkg/mod/gioui.org@v0.1.0/layout/flex.go:99 +0x2d5
github.com/crypto-power/cryptopower/ui/page/components.(*SubPage).Layout(0xc008b99c70, {0x1f06330?, 0xc0005303f0}, {{{0x0, 0x0}, {0x550, 0x3e7}}, {0x3f800000, 0x3f800000}, {0x1ef1be0, ...}, ...})
        /home/kennedy/Projects/src/github.com/cryptopower/cryptopower/ui/page/components/sub_page.go:43 +0x1d4
github.com/crypto-power/cryptopower/ui/page/exchange.(*OrderDetailsPage).Layout.func1({{{0x0, 0x0}, {0x550, 0x3e7}}, {0x3f800000, 0x3f800000}, {0x1ef1be0, 0xc0038780c8}, {0xc15051ead6a83efa, 0x23eb3e0b5a, ...}, ...})
        /home/kennedy/Projects/src/github.com/cryptopower/cryptopower/ui/page/exchange/order_details_page.go:136 +0x286
gioui.org/layout.Inset.Layout({0x2?, 0x0?, 0x0?, 0x0?}, {{{0x0, 0x0}, {0x550, 0x3e7}}, {0x3f800000, 0x3f800000}, ...}, ...)
        /home/kennedy/Projects/pkg/mod/gioui.org@v0.1.0/layout/layout.go:169 +0x39d
github.com/crypto-power/cryptopower/ui/page/components.UniformPadding({{{0x0, 0x0}, {0x780, 0x417}}, {0x3f800000, 0x3f800000}, {0x1ef1be0, 0xc0038780c8}, {0xc15051ead6a83efa, 0x23eb3e0b5a, ...}, ...}, ...)
        /home/kennedy/Projects/src/github.com/cryptopower/cryptopower/ui/page/components/components.go:96 +0x197
github.com/crypto-power/cryptopower/ui/page/exchange.(*OrderDetailsPage).Layout(0x1ff7?, {{{0x0, 0x0}, {0x780, 0x417}}, {0x3f800000, 0x3f800000}, {0x1ef1be0, 0xc0038780c8}, {0xc15051ead6a83efa, ...}, ...})
        /home/kennedy/Projects/src/github.com/cryptopower/cryptopower/ui/page/exchange/order_details_page.go:139 +0x65
github.com/crypto-power/cryptopower/ui.(*Window).prepareToDisplayUI.func2({{{0x0, 0x0}, {0x780, 0x417}}, {0x3f800000, 0x3f800000}, {0x1ef1be0, 0xc0038780c8}, {0xc15051ead6a83efa, 0x23eb3e0b5a, ...}, ...})
        /home/kennedy/Projects/src/github.com/cryptopower/cryptopower/ui/window.go:282 +0x122
gioui.org/layout.Stack.Layout({0x20?}, {{{0x780, 0x417}, {0x780, 0x417}}, {0x3f800000, 0x3f800000}, {0x1ef1be0, 0xc0038780c8}, {0xc15051ead6a83efa, ...}, ...}, ...)
        /home/kennedy/Projects/pkg/mod/gioui.org@v0.1.0/layout/stack.go:60 +0x195
github.com/crypto-power/cryptopower/ui.(*Window).prepareToDisplayUI(0xc000540510, {{0xc15051ead6a83efa, 0x23eb3e0b5a, 0x2e13620}, {0x3f800000, 0x3f800000}, {0x780, 0x417}, {0x0, 0x0, ...}, ...})
        /home/kennedy/Projects/src/github.com/cryptopower/cryptopower/ui/window.go:297 +0x40d
github.com/crypto-power/cryptopower/ui.(*Window).handleFrameEvent(0xc000540510, {{0xc15051ead6a83efa, 0x23eb3e0b5a, 0x2e13620}, {0x3f800000, 0x3f800000}, {0x780, 0x417}, {0x0, 0x0, ...}, ...})
        /home/kennedy/Projects/src/github.com/cryptopower/cryptopower/ui/window.go:235 +0x225
github.com/crypto-power/cryptopower/ui.(*Window).HandleEvents(0xc000540510)
        /home/kennedy/Projects/src/github.com/cryptopower/cryptopower/ui/window.go:200 +0x2b9
main.main.func3()
        /home/kennedy/Projects/src/github.com/cryptopower/cryptopower/main.go:109 +0x1d
created by main.main
        /home/kennedy/Projects/src/github.com/cryptopower/cryptopower/main.go:107 +0x51d
exit status 2
```

**Comment by dreacot** [comment]
> trying to swap ltc and dcr throws an error Untitled

speaking to the trocador team, this error occurs because the miminum trade limit is not met, they have also agreed to look into improving the error message.

the minimum returned from the server doesn't seem to be accurate, we might need to override that with sane defaults for trocador

**Comment by vibros68** [comment]
> > trying to swap ltc and dcr throws an error Untitled
> 
> speaking to the trocador team, this error occurs because the miminum trade limit is not met, they have also agreed to look into improving the error message.
> 
> the minimum returned from the server doesn't seem to be accurate, we might need to override that with sane defaults for trocador

The response from trocador is not enough for us to handle this error. It should be fixed on trocador side

**Comment by dreacot** [comment]
> > > trying to swap ltc and dcr throws an error Untitled
> > 
> > 
> > speaking to the trocador team, this error occurs because the miminum trade limit is not met, they have also agreed to look into improving the error message.
> > the minimum returned from the server doesn't seem to be accurate, we might need to override that with sane defaults for trocador
> 
> The response from trocador is not enough for us to handle this error. It should be fixed on trocador side

can we override the minimum trade amount from the server? to prevent users from trading too low

**Review by dreacot — COMMENTED** [review]
Looks good, i think insantswap version needs to be updated

**Review by dreacot — APPROVED** [review]

File: ui/page/exchange/create_order_page.go [hunk 3]
```diff
@@ -1318,3 +1334,12 @@ func (pg *CreateOrderPage) listenForNotifications() {
 func (pg *CreateOrderPage) stopNtfnListeners() {
 	pg.WL.AssetsManager.InstantSwap.RemoveNotificationListener(CreateOrderPageID)
 }
+
+func (pg *CreateOrderPage) getInstantCurrencyInfos() error {
```
Comment (ukane-philemon) [inline ui/page/exchange/create_order_page.go:1338]: Rename. Maybe `instantExchangeCurrencyInfos`.

File: ui/page/exchange/create_order_page.go [hunk 4]
```diff
@@ -1318,3 +1334,12 @@ func (pg *CreateOrderPage) listenForNotifications() {
...
+	pg.fetchingRate = true
+	currencies, err := pg.exchange.GetCurrencies()
+	pg.instantCurrencies = currencies
```
Comment (ukane-philemon) [inline ui/page/exchange/create_order_page.go:1341]: Please rename this property(`pg.instantCurrencies`).

File: ui/page/exchange/order_scheduler_modal.go [hunk 5]
```diff
@@ -492,26 +503,64 @@ func (osm *orderSchedulerModal) startOrderScheduler() {
 	}()
 }
 
+func (osm *orderSchedulerModal) getInstantCurrencyInfos() error {
```
Comment (ukane-philemon) [inline ui/page/exchange/order_scheduler_modal.go:506]: Pls rename. Maybe `instantExchangeCurrencyInfos`.

**Review by ukane-philemon — CHANGES_REQUESTED** [review]
Looks good. First pass.

File: .github/workflows/go.yml [hunk 6]
```diff
@@ -19,6 +19,10 @@ jobs:
           sudo apt update
           sudo apt install libwayland-dev libx11-dev libx11-xcb-dev libxkbcommon-x11-dev libgles2-mesa-dev libegl1-mesa-dev libffi-dev libxcursor-dev libvulkan-dev
 
+      - name: Copy needed files
+        run: |
+          cp "./libwallet/instantswap/instant_example.json" "./libwallet/instantswap/instant.json"
+
```
[the remark this answers is not in the document]
Comment (dreacot) [reply .github/workflows/go.yml:25]: This is the workflow file

File: libwallet/instantswap/instantswap.go [hunk 7]
```diff
@@ -11,23 +13,51 @@ import (
...
+	}
+	// delete instant exchange with no key
+	for key, val := range privKeyMap {
+		if val == "" {
+			delete(privKeyMap, key)
+		}
+	}
+	// add flypme to privKeyMap because it does not requires private key to access
+	privKeyMap[FlypMe] = ""
+}
```
[the remark this answers is not in the document]
Comment (vibros68) [reply libwallet/instantswap/instantswap.go:55]: privKeyMap is read-only. It is initial at the compiler time. So add sync.Map is not needed. Make the code redundant lengthy

File: libwallet/instantswap/instantswap.go [hunk 8]
```diff
@@ -154,19 +184,22 @@ func (instantSwap *InstantSwap) GetOrderByIDRaw(orderID int) (*Order, error) {
 }
 
 func (instantSwap *InstantSwap) CreateOrder(exchangeObject instantswap.IDExchange, params Order) (*Order, error) {
-	const op errors.Op = "instantSwap.CreateOrder"
```
[the remark this answers is not in the document]
Comment (vibros68) [reply libwallet/instantswap/instantswap.go:157]: Display error op in the user interface is not friendly. It should put on error log

File: ui/page/exchange/confirm_order_modal.go [hunk 9]
```diff
@@ -124,7 +125,7 @@ func (com *confirmOrderModal) confirmOrder() {
 
 		order, err := com.createOrder()
 		if err != nil {
-			log.Error(err)
+			log.Error(errors.E(errors.Op("instantSwap.CreateOrder"), err))
```
[the remark this answers is not in the document]
Comment (vibros68) [reply ui/page/exchange/confirm_order_modal.go:128]: As explained above. error op is hided from ui

File: libwallet/instantswap/instantswap.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (ukane-philemon) [reply libwallet/instantswap/instantswap.go:55]: > privKeyMap is read-only.

Ok.

**Comment by ukane-philemon** [comment]
> I signed up to get an API key from Trocador but didn't get one. I just can't locate where I'm supposed to copy the API key.
> 
> utAck, just these changes and it's GTG.

@vibros68, where can I find the API key on the Trocador website? a screenshot, please.

**Comment by ukane-philemon** [comment]
> You have to contact their supporter via mail.

omo, interesting 🤔.

**Comment by ukane-philemon** [comment]
@dreacot I  see this but dunno if we should open a separate PR for it. The conversion from Flypme is way off, I think.

Screenshot 2023-12-01 at 6 04 48 PM

**Comment by dreacot** [comment]
> @dreacot I see this but dunno if we should open a separate PR for it. The conversion from Flypme is way off, I think.
> 
> Screenshot 2023-12-01 at 6 04 48 PM

good catch!. is it on master? if it is, then open a separate issue

**Comment by ukane-philemon** [comment]
> good catch!. is it on master?

No, it's not.

**Comment by dreacot** [comment]
> > good catch!. is it on master?
> 
> No, it's not.
> 
> master: Screenshot 2023-12-01 at 6 12 24 PM

@vibros68 please look into the rate conversion

**Comment by vibros68** [comment]
> > > good catch!. is it on master?
> > 
> > 
> > No, it's not.
> > master: Screenshot 2023-12-01 at 6 12 24 PM
> 
> @vibros68 please look into the rate conversion

after investigate, I find out that it is a bug from instantswap. It canculates finaRate = 1/exchangeRate from the old exchange(flypme,changenow,godex,...) While trocador is not. In the old code exchangeRate is taken by EstimatedAmount field with FromAmount = 1. Generally we have to fix it from instantswap

**Comment by dreacot** [comment]
> > > > good catch!. is it on master?
> > > 
> > > 
> > > No, it's not.
> > > master: Screenshot 2023-12-01 at 6 12 24 PM
> > 
> > 
> > @vibros68 please look into the rate conversion
> 
> after investigate, I find out that it is a bug from instantswap. It canculates finaRate = 1/exchangeRate from the old exchange(flypme,changenow,godex,...) While trocador is not. In the old code exchangeRate is taken by EstimatedAmount field with FromAmount = 1. Generally we have to fix it from instantswap

File: go.mod [hunk 10]
```diff
@@ -156,3 +156,5 @@ require (
 // the android OS using gomobile. This replace can be removed once any
 // of those projects update their github.com/lib/pq dependency.
 replace github.com/lib/pq => github.com/lib/pq v1.10.4
+
+replace github.com/crypto-power/instantswap => github.com/vibros68/instantswap v0.0.0-20231205093815-045a3f9f0a0c
```
Comment (ukane-philemon) [inline go.mod:160]: No, submit a PR to instantswap.

Comment (vibros68) [reply go.mod:160]: It is for reviewing. After the review is done, we will merge on both instantswap and cryptopower.
@dreacot please confirm on it

Comment (ukane-philemon) [reply go.mod:160]: Okay, I’ve seen your instantswap PR.

**Review by dreacot — APPROVED** [review]

**Pull request merged**